wp_api/src/wordpress_org/client.rs

```diff
@@ -0,0 +1,252 @@
+use crate::{
+    request::{endpoint::WpEndpointUrl, RequestExecutor, WpNetworkRequest, WpNetworkResponse},
+    RequestExecutionError,
+};
+use serde::de::DeserializeOwned;
+use std::{result::Result, sync::Arc};
+use url::Url;
+
+use super::plugin_directory::{PluginInformation, QueryPluginResponse};
+
+#[derive(Debug, uniffi::Object)]
+pub struct WordPressOrgApiClient {
+    pub(crate) request_executor: Arc<dyn RequestExecutor>,
+}
+
+#[uniffi::export]
+impl WordPressOrgApiClient {
+    #[uniffi::constructor]
+    pub fn new(request_executor: Arc<dyn RequestExecutor>) -> Self {
+        Self { request_executor }
+    }
+
+    pub async fn plugin_information(
+        &self,
+        slug: &str,
+    ) -> Result<PluginInformation, WordPressOrgApiClientError> {
+        self.execute(self.plugin_information_request(slug)).await
+    }
+
+    pub fn browse_plugins_request(
+        &self,
```

Comment:
We don't need to take `&self` for building requests. Once it's removed, we also won't require a mock executor in the unit tests.

Uniffi won't allow us to have associated functions, so they'd need to be moved to the unexported `impl WordPressOrgApiClient` block which we have 2 of at the moment - and I think we should stick to just one in this case. This means native wrappers won't be able to access the build request functions, but if that ends up being necessary, we can extract them into a `RequestBuilder` type which is why we have that abstraction for all other clients. For now, we can skip this step.

This suggestion requires moving a few pieces around which I already did to make sure it all works. If you like, you can apply the following diff:

```
diff --git a/wp_api/src/wordpress_org/client.rs b/wp_api/src/wordpress_org/client.rs
index fc8f419..1796bc9 100644
--- a/wp_api/src/wordpress_org/client.rs
+++ b/wp_api/src/wordpress_org/client.rs
@@ -24,62 +24,56 @@ impl WordPressOrgApiClient {
         &self,
         slug: &str,
     ) -> Result<PluginInformation, WordPressOrgApiClientError> {
-        self.execute(self.plugin_information_request(slug)).await
+        self.execute(Self::plugin_information_request(slug)).await
     }
 
-    pub fn browse_plugins_request(
+    pub async fn browse_plugins(
         &self,
         category: Option<WordPressOrgApiPluginDirectoryCategory>,
         page: u64,
         page_size: u64,
-    ) -> WpNetworkRequest {
-        self.query_plugins_request(page, page_size, |url| match category {
-            Some(category) => {
-                let mut url = url;
-                url.query_pairs_mut()
-                    .append_pair("browse", category.as_str());
-                url
-            }
-            None => url,
-        })
+    ) -> Result<QueryPluginResponse, WordPressOrgApiClientError> {
+        let request = Self::browse_plugins_request(category, page, page_size);
+        self.execute(request).await
     }
 
-    pub async fn browse_plugins(
+    pub async fn search_plugins(
         &self,
-        category: Option<WordPressOrgApiPluginDirectoryCategory>,
+        search: String,
         page: u64,
         page_size: u64,
     ) -> Result<QueryPluginResponse, WordPressOrgApiClientError> {
-        let request = self.browse_plugins_request(category, page, page_size);
+        let request = Self::search_plugins_request(search, page, page_size);
         self.execute(request).await
     }
+}
 
-    pub fn search_plugins_request(
-        &self,
-        search: String,
+impl WordPressOrgApiClient {
+    pub fn browse_plugins_request(
+        category: Option<WordPressOrgApiPluginDirectoryCategory>,
         page: u64,
         page_size: u64,
     ) -> WpNetworkRequest {
-        self.query_plugins_request(page, page_size, |url| {
+        Self::query_plugins_request(page, page_size, |url| match category {
+            Some(category) => {
+                let mut url = url;
+                url.query_pairs_mut()
+                    .append_pair("browse", category.as_str());
+                url
+            }
+            None => url,
+        })
+    }
+
+    pub fn search_plugins_request(search: String, page: u64, page_size: u64) -> WpNetworkRequest {
+        Self::query_plugins_request(page, page_size, |url| {
             let mut url = url;
             url.query_pairs_mut().append_pair("search", &search);
             url
         })
     }
 
-    pub async fn search_plugins(
-        &self,
-        search: String,
-        page: u64,
-        page_size: u64,
-    ) -> Result<QueryPluginResponse, WordPressOrgApiClientError> {
-        let request = self.search_plugins_request(search, page, page_size);
-        self.execute(request).await
-    }
-}
-
-impl WordPressOrgApiClient {
-    fn plugin_information_request(&self, slug: &str) -> WpNetworkRequest {
+    fn plugin_information_request(slug: &str) -> WpNetworkRequest {
         let mut url = Self::plugin_info_api_url();
         url.query_pairs_mut()
             .append_pair("action", "plugin_information")
@@ -88,12 +82,7 @@ impl WordPressOrgApiClient {
         WpNetworkRequest::get(WpEndpointUrl(url.to_string()))
     }
 
-    fn query_plugins_request<F>(
-        &self,
-        page: u64,
-        page_size: u64,
-        url_builder: F,
-    ) -> WpNetworkRequest
+    fn query_plugins_request<F>(page: u64, page_size: u64, url_builder: F) -> WpNetworkRequest
     where
         F: FnOnce(Url) -> Url,
     {
@@ -106,6 +95,10 @@ impl WordPressOrgApiClient {
         WpNetworkRequest::get(WpEndpointUrl(url.to_string()))
     }
 
+    fn plugin_info_api_url() -> Url {
+        Url::parse("https://api.wordpress.org/plugins/info/1.2/").expect("The URL is valid")
+    }
+
     async fn execute<T>(&self, request: WpNetworkRequest) -> Result<T, WordPressOrgApiClientError>
     where
         T: DeserializeOwned,
@@ -113,12 +106,6 @@ impl WordPressOrgApiClient {
         let response = self.request_executor.execute(Arc::new(request)).await?;
         Self::parse(response)
     }
-}
-
-impl WordPressOrgApiClient {
-    fn plugin_info_api_url() -> Url {
-        Url::parse("https://api.wordpress.org/plugins/info/1.2/").expect("The URL is valid")
-    }
 
     fn parse<T>(response: WpNetworkResponse) -> Result<T, WordPressOrgApiClientError>
     where
@@ -195,52 +182,17 @@ impl From<RequestExecutionError> for WordPressOrgApiClientError {
 
 #[cfg(test)]
 mod tests {
-
     use super::*;
-    use crate::{
-        request::{endpoint::media_endpoint::MediaUploadRequest, RequestExecutor},
-        MediaUploadRequestExecutionError,
-    };
-    use std::sync::Arc;
-
-    #[derive(Debug)]
-    struct MockRequestExecutor;
-
-    #[async_trait::async_trait]
-    impl RequestExecutor for MockRequestExecutor {
-        async fn execute(
-            &self,
-            _wp_request: Arc<WpNetworkRequest>,
-        ) -> std::result::Result<WpNetworkResponse, RequestExecutionError> {
-            Err(RequestExecutionError::RequestExecutionFailed {
-                status_code: None,
-                reason: "Mocked request executor".to_string(),
-            })
-        }
-
-        async fn upload_media(
-            &self,
-            media_upload_request: Arc<MediaUploadRequest>,
-        ) -> std::result::Result<WpNetworkResponse, MediaUploadRequestExecutionError> {
-            unimplemented!(
-                "upload_media is not implemented for sending requests to api.wordpress.org"
-            )
-        }
-    }
 
     #[test]
     fn test_plugin_info_requests_include_icons() {
-        let request_executor = Arc::new(MockRequestExecutor {});
-        let client = WordPressOrgApiClient::new(request_executor.clone());
-        let request = client.plugin_information_request("akismet");
+        let request = WordPressOrgApiClient::plugin_information_request("akismet");
         assert!(request.url.0.contains("fields=icons"));
     }
 
     #[test]
     fn test_search_does_not_include_pagination() {
-        let request_executor = Arc::new(MockRequestExecutor {});
-        let client = WordPressOrgApiClient::new(request_executor.clone());
-        let request = client.search_plugins_request("akismet".to_string(), 3, 24);
+        let request = WordPressOrgApiClient::search_plugins_request("akismet".to_string(), 3, 24);
 
         // The 'request[x]' parameters do not work for the search endpoint.
         // The 'page' and 'per_page' parameters do.
```

Reply:
⬆️ Diff is taken in https://github.com/Automattic/wordpress-rs/pull/445/commits/d2ebf6e0aa75278b2e43c0ff73da0cc5b8ebe310